### DISCUSSION TITLE
Clink in VSCode integrated terminal: CTRL+SPACE/Autocompletion doesn't work

### DISCUSSION THREAD

**Discussion by sewbacca** [chrisant996/clink#423] [Q&A] [answered]
Hello folks,
I want to use Clink inside the integrated terminal of VSCode, but autocompletion does not work correctly:

When hitting `CTRL+SPACE` nothing happens. Tab completion works as expected (I am using bash layout).
`CTRL+SPACE` works inside a cmd window and the new Windows Terminal. When I check what is received with `SHIFT+ALT+?` then I get:
```
C-@ : set-mark -- Set the mark to the cursor point.  If a numeric argument is supplied, sets the mark to that position
```
while in a cmd console window I get:
```
C-Space : clink-select-complete -- Perform completion by selecting from an interactive list of possible completions;
if there is only one match, insert it
```
`C-@` is also received when I manually send `\u0000` to the console:
```json
{
	"key": "ctrl+space",
	"when": "terminalFocus",
	"command": "workbench.action.terminal.sendSequence",
	"args": {
		"text": "\u0000",
	}
}
```
So VSCode is not capturing `CTRL+SPACE` during `terminalFocus`

Any help is appreciated

**Comment by chrisant996** [comment]
It's a VSCode problem.  Or at least, it's not caused by Clink.

You can use `clink echo --verbose` to see how keyboard input is received by Clink.

Outside of VSCode it reports the input as:
```plain
key event:  _C_ __  flags=0x00000008  char=0x0020  vk=0x0020  scan=0x0039  "Space"
"\e[27;5;32~"
```
- The `_C` means Ctrl is held.
- The `0x0020` means Space was pressed.

Inside of VSCode it reports the input as:
```plain
key event:  SC_ __  flags=0x00000018  char=0x0000  vk=0x0032  scan=0x0003  "2"
"\C-@"
```
- The `SC` means Shift and Ctrl are held -- which is inaccurate.
- The `0x0032` means the 2 key was pressed -- which is inaccurate.

Something upstream from Clink is forwarding inaccurate keyboard input to the embedded terminal inside VSCode.  Probably it is going wrong inside VSCode's embedded terminal support.  But it could potentially be going wrong somewhere else.

**Comment by chrisant996** [comment, marked as answer]
You could report the problem in the VSCode repo, or you could live with it and set a key binding for `"\C-@": clink-select-complete` in your `.inputrc` file.

**Reply by sewbacca** [reply]
Thanks for the workaround. I have created an [issue](https://github.com/microsoft/vscode/issues/174618) in the VSCode repo. I'll get back to you once I've got an answer.

**Reply by sewbacca** [reply]
Seems to be an [issue with xterm](https://github.com/xtermjs/xterm.js/issues/2357) and VSCode tries it best to remap certain keystrokes for PowerShell.
